src/charm.py

```diff
@@ -125,6 +157,48 @@ def __init__(self, *args):
         # Action handlers
         observe(self.on.show_proxied_endpoints_action, self._on_show_proxied_endpoints)
 
+    def _on_install(self, event) -> None:
+        private_key_password = self._generate_password()
+        private_key = generate_private_key(password=private_key_password.encode("utf-8"))
+        self._stored.private_key_password = private_key_password
+        self._stored.private_key = private_key.decode()
+
+    def _on_certificates_relation_joined(self, event: RelationJoinedEvent) -> None:
+        private_key_password = self._stored.private_key_password
+        private_key = self._stored.private_key
+        csr = generate_csr(
+            private_key=private_key.encode("utf-8"),
+            private_key_password=private_key_password.encode("utf-8"),
+            subject=self.cert_subject,
+        )
+        self._stored.csr = csr.decode()
+        self.certificates.request_certificate_creation(certificate_signing_request=csr)
+
+    def _on_certificate_available(self, event: CertificateAvailableEvent) -> None:
+        self._stored.certificate = event.certificate
+        self._stored.ca = event.ca
+        self._stored.chain = event.chain
+        # TODO: Store files in container and modify config file
+        self.container.push(_CERTIFICATE_PATH, self._stored.certificate, make_dirs=True)
+        self.container.push(_CERTIFICATE_KEY_PATH, self._stored.private_key, make_dirs=True)
```

Comment:
Possibly/probably yes. This is more of a PM question, but considering the essentials of certs and subdomains (n >= 2 total domains/subdomains requires multiple certs), it's probably a safe bet to design it around that assumption, make `self._stored....` values for certs into dicts, and drag `_type_convert_stored` in here, too and start from a position of flexibilty rather than needing to try to potentially migrate it later.

Reply:
(I forgot to resolve this comment.)
Iirc we discussed this a while ago and decided there will be only one and also added `llimit: 1` to metadata.
Afaiu, Traefik's `external_host` is just one host, and any subdomains could be added to the same CSR.
Am I missing anything here?

Reply:
The real question being that, like an earlier discussion, you can't have a wildcard for `*.example.org` which also works for `*.foo.example.org`, and they can't be part of the same CSR even with a wildcard. 

We also decided on "one Traefik per subdomain for now, but we may as well make it a little flexible just in case that needs to change so we don't have to break API if it comes up later".

For safety, a comma-separated list of domains (which could be merged with the domain of `external_host`) which are kept in storedstate, and making `_stored.private_key | certificate` into lists/dicts would possibly save a lot of potential pain in the future.